src/backend/storage/course_store.rs

```diff
@@ -1,42 +1,122 @@
-                    "src/backend/storage/sample_queries/select_courses.sql",
-                    ["%", course_name, "%"].concat()
-                )
-                .fetch_all(&self.pool)
-                .await
-            }
-        }
+    /// Selects courses for the current year from a pre-built materialized view.
+    pub async fn select_courses(&self, course_code: &str) -> Result<Vec<Course>, StorageError> {
+        let sql = "
+        SELECT
+            id,
+            catalog_number,
+            subject_code,
+            title,
+            external_id,
+            offerings
+        FROM
+            mv_courses
+        WHERE
+            subject_code || catalog_number::VARCHAR ILIKE $1
+        GROUP BY
+            id,
+            title,
+            external_id,
+            subject_code,
+            catalog_number,
+            offerings
+        ORDER BY
+            catalog_number
+        LIMIT
+            100;
+        ";
+
+        sqlx::query_as::<Postgres, Course>(sql)
+            .bind(["%", &course_code.to_uppercase(), "%"].concat())
+            .fetch_all(&self.pool)
+            .await
+            .map_err(StorageError::QueryFailure)
+    }
+
+    pub async fn select_course_by_code(&self, course_code: &str) -> Result<Course, StorageError> {
+        let sql = "
+        SELECT
+            c.id AS id,
+            c.catalog_number AS catalog_number,
+            c.subject_code AS subject_code,
+            c.external_id AS external_id,
+            c.academic_level AS academic_level,
+            c.title AS title,
+            c.description AS description,
+            COALESCE(
+                NULLIF(json_agg(DISTINCT co.*)::TEXT, '[null]'),
+                '[]'
+            ) AS offerings,
+            COALESCE(
+                NULLIF(array_agg(DISTINCT rc.subject_code || rc.catalog_number::TEXT), '{null}'),
+                '{}'
+            ) AS required_prerequisites,
+            COALESCE(
+                NULLIF(array_agg(DISTINCT oc.subject_code || oc.catalog_number::TEXT), '{null}'),
+                '{}'
+            ) AS optional_prerequisites
+        FROM
+            courses c
+            LEFT JOIN course_offerings co ON c.id = co.course_id
+            LEFT JOIN prerequisites p ON c.prerequisites_id = p.id
+            LEFT JOIN required_prerequisites rp ON p.id = rp.prerequisite_id
+            LEFT JOIN courses rc ON rp.course_id = rc.id
+            LEFT JOIN optional_prerequisites op ON p.id = op.prerequisite_id
+            LEFT JOIN courses oc ON op.course_id = oc.id
+        WHERE
+            c.subject_code || c.catalog_number::VARCHAR = $1
+        GROUP BY
+            c.id
+        LIMIT 1;
+        ";
+
+        sqlx::query_as::<Postgres, Course>(sql)
+            .bind(course_code.to_uppercase())
+            .fetch_one(&self.pool)
+            .await
+            .map_err(StorageError::MissingRecords)
+
+        // TODO: enable conditional sql query usage
```

Comment:
IMO it seems a bit unnecessary to branch on production or staging on each query. I'd suggest branching and setting a single file path in the constructor?

Reply:
That's fair. I think we should aim for all three of these:
1. Using sqlx macros unless the builder is absolutely necessary. This ensures that our queries are always valid and helps catch errors at compile time. If we change our DDL next week but forget to update a query, the app will not compile. This is a great feature and has already saved us hours of debugging.
2. Modifying the .env file should allow us to switch between production and sample storage modes.
3. The storage mode doesn't affect the rest of the app.

If we want all three, the only other option I can think of is defining a Trait for a storage class, defining a Sample and Production implementation of the Trait, and making our Storage class generic over the trait. This is essentially branching during construction. I'm alright with anything as long as we have all three.

Reply:
I mean we can simply determine the file name of the sql queries to use in the constructor based on sample or prod env var.

But for the first milestone requirements, I still don't see the value in interoping with sample or prod queries in the code itself since they produce the same rows. If we want to performance tune, we can just use EXPLAIN ANALYZE.

Reply:
As we discussed for #9, that isn't possible with macros as each path needs to be a string literal.

This isn't necessary for the first milestone, but we merged #29 and closed #9 two days ago. This is a regression.

Reply:
I don’t really think we should have a macro to choose between prod or sample sql queries since the entire data model and normalization structure of the tables could totally change.

Part of optimizing the query also involves in optimizing the layout of the data too, not just how it’s selected. If we denormalize tables to reduce joins, the mapping from rows to rust struct could totally change. Having to do this multiple ways in application code overcomplicates things.